File: Wallet.php

class Wallet {
    private $balance;
    private $username;
    private $password;

    public function __construct($username, $password) {
        $this->balance = 0;
        $this->username = $username;
        $this->password = $password;
    }

    public function deposit($amount) {
        $this->authenticate();
        $this->balance += $amount;
        echo "Deposit successful\n";
    }

    public function withdraw($amount) {
        $this->authenticate();
        if ($amount <= $this->balance) {
            $this->balance -= $amount;
            echo "Withdrawal successful\n";
        } else {
            echo "Insufficient funds\n";
        }
    }

    public function getBalance() {
        $this->authenticate();
        return $this->balance;
    }

    private function authenticate() {    
        echo "Enter username: ";
        $inputUsername = trim(fgets(STDIN));
        echo "Enter password: ";
        $inputPassword = trim(fgets(STDIN));

        if ($this->username !== $inputUsername || $this->password !== $inputPassword) {
            echo "Authentication failed. Access denied.\n";
            exit(1); 
        }
    }
}

File: Wallet.py

class Portfolio:
    def __init__(self, username, password):
        self.balance = 0
        self.username = username
        self.password = password
        self.transactions = []

    def deposit(self, amount):
        if amount > 0:
            self.balance += amount
            self.transactions.append(f"Deposit: +{amount}")

    def withdraw(self, amount):
        if amount > 0 and amount <= self.balance:
            self.balance -= amount
            self.transactions.append(f"Withdrawal: -{amount}")
        else:
            print("Insufficient funds or invalid withdrawal amount.")

    def display_balance(self):
        print(f"Current balance: {self.balance}")

    def authenticate(self, input_username, input_password):
        return input_username == self.username and input_password == self.password

    def display_transactions(self):
        print("Transaction history:")
        for transaction in self.transactions:
            print(transaction)
